File: sdk/typescript/src/manifest.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { ManifestMismatchError } from "./errors.js";
import type { Encoding } from "./encode.js";

export const DATA_DIR = fileURLToPath(new URL("../data/", import.meta.url));
const SUPPORTED_MANIFEST_VERSION = 1;

export interface Binding { path: string; encode: Encoding }
export interface AttributePlan { field: string; unit: string | null; quantity_kind: string | null }
export interface ComponentPlan {
  typeName: string;
  rank: number;
  entity_sql: string;
  row_sql: string;
  bindings: Binding[];
  entityBindings: Binding[];
  attributes: AttributePlan[];
  gaps: string[];
  knownFields: Set<string>;
}
export interface AssociationPlan { section: string; row_sql: string; bindings: Binding[] }
export interface Manifest {
  schema_user_version: number;
  vocabulary: {
    entity_types: { name: string; is_topology: boolean; is_dc: boolean }[];
    [table: string]: unknown;
  };
  components: Record<string, ComponentPlan>;
  unsupported_components: Record<string, string>;
  attribute_sql: string;
  supplemental_attributes: { plant_types: string[]; plant_sql: string; attribute_sql: string };
  associations: AssociationPlan[];
  unsupported_sections: Record<string, string>;
}

let cached: Manifest | undefined;

export function dataText(name: string): string {
  return readFileSync(DATA_DIR + name, "utf-8");
}

export function loadManifest(): Manifest {
  if (cached) return cached;
  const raw = JSON.parse(dataText("insert_manifest.json"));
  if (raw.manifest_version !== SUPPORTED_MANIFEST_VERSION) {
    throw new ManifestMismatchError(
      `manifest_version ${raw.manifest_version} is not ${SUPPORTED_MANIFEST_VERSION}`,
    );
  }
  const components: Record<string, ComponentPlan> = {};
  for (const [typeName, entry] of Object.entries<any>(raw.components)) {
    const knownFields = new Set<string>([
      ...entry.bindings.map((b: Binding) => b.path.split(".")[0]),
      ...entry.attributes.map((a: AttributePlan) => a.field),
      ...entry.skip,
      ...entry.gaps,
    ]);
    components[typeName] = {
      ...entry,
      typeName,
      entityBindings: entry.bindings.slice(0, 1),
      knownFields,
    };
  }
  cached = { ...raw, components } as Manifest;
  return cached;
}

File: sdk/julia/src/manifest.jl
const SUPPORTED_MANIFEST_VERSION = 1

struct Binding
    segments::Vector{String}
    encoding::Encoding
end

struct AttributePlan
    field::String
    registered::Bool
    unit::String
    quantity_kind::String
end

struct ComponentPlan
    type_name::String
    rank::Int
    entity_sql::String
    row_sql::String
    bindings::Vector{Binding}
    entity_bindings::Vector{Binding}
    attributes::Vector{AttributePlan}
    gaps::Vector{String}
    known_fields::Set{String}
end

struct AssociationPlan
    section::String
    row_sql::String
    bindings::Vector{Binding}
end

struct Manifest
    schema_user_version::Int
    vocabulary::Dict{String, Any}
    components::Dict{String, ComponentPlan}
    unsupported_components::Dict{String, String}
    attribute_sql::String
    plant_types::Set{String}
    plant_sql::String
    supplemental_attribute_sql::String
    associations::Vector{AssociationPlan}
    unsupported_sections::Dict{String, String}
end

parse_bindings(raw) =
    Binding[Binding(String.(split(b["path"], '.')), ENCODINGS[b["encode"]]) for b in raw]

parse_attribute(raw, ::Nothing) = AttributePlan(raw["field"], false, "", "")
parse_attribute(raw, unit::AbstractString) =
    AttributePlan(raw["field"], true, unit, raw["quantity_kind"])

function parse_component(type_name::String, raw::AbstractDict)
    bindings = parse_bindings(raw["bindings"])
    attributes = AttributePlan[parse_attribute(a, a["unit"]) for a in raw["attributes"]]
    known = Set{String}()
    for b in bindings
        push!(known, first(b.segments))
    end
    for a in attributes
        push!(known, a.field)
    end
    union!(known, String.(raw["skip"]), String.(raw["gaps"]))
    return ComponentPlan(
        type_name,
        raw["rank"],
        raw["entity_sql"],
        raw["row_sql"],
        bindings,
        bindings[1:1],
        attributes,
        String.(raw["gaps"]),
        known,
    )
end

function parse_manifest(raw::AbstractDict)
    if raw["manifest_version"] != SUPPORTED_MANIFEST_VERSION
        throw(
            ManifestMismatchError(
                "manifest_version $(raw["manifest_version"]) is not $SUPPORTED_MANIFEST_VERSION",
            ),
        )
    end
    components = Dict{String, ComponentPlan}(
        String(n) => parse_component(String(n), e) for (n, e) in raw["components"]
    )
    supplemental = raw["supplemental_attributes"]
    associations = AssociationPlan[
        AssociationPlan(a["section"], a["row_sql"], parse_bindings(a["bindings"])) for
        a in raw["associations"]
    ]
    return Manifest(
        raw["schema_user_version"],
        raw["vocabulary"],
        components,
        Dict{String, String}(raw["unsupported_components"]),
        raw["attribute_sql"],
        Set{String}(supplemental["plant_types"]),
        supplemental["plant_sql"],
        supplemental["attribute_sql"],
        associations,
        Dict{String, String}(raw["unsupported_sections"]),
    )
end

const MANIFEST = Ref{Manifest}()

function manifest()
    if !isassigned(MANIFEST)
        path = joinpath(DATA_DIR, "insert_manifest.json")
        MANIFEST[] = parse_manifest(JSON.parsefile(path; dicttype=Dict{String, Any}))
    end
    return MANIFEST[]
end
